### PR TITLE
Bluetooth: Host: Add LE Power Control Procedure APIs

### PR DESCRIPTION
**Pull request by KyraLengfeld**
This PR aims to define host APIs, events and configs for LE Power Control Feature. The support of feature is provided with controller-based feature selection with BT_CTLR_LE_POWER_CONTROL_SUPPORT. If the feature is supported, then BT_TRANSMIT_POWER_CONTROL feature will be selectable.

With the new APIs, the applications will:

get improved reading of local and remote tx power
be aware of changes in remote or local tx power
be able to request a tx power change on remote device
be able to set automatic Power Control Request when leaving golden range of RSSI
be able to set APR handling of remote apr to use in local tx power setting
read average RSSI

Defined HCI commands in Core Spec v5.3:

7.8.117 LE Enhanced Read Transmit Power Level command: improvement to existing local tx power reading.
7.8.118 LE Read Remote Transmit Power Level command: Remote tx power is read through an event (LE Transmit Power Reporting)
7.8.121 LE Set Transmit Power Reporting Enable command: Enables local or remote tx power reporting to monitor changes in tx power
7.7.65.33 LE Transmit Power Reporting event:

Vendor-specific HCI commands:
Write remote transmit power level
Set LE Power Control Request parameters

### CODE REVIEW

File: include/zephyr/bluetooth/conn.h [hunk 1]
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
...
+					  bool local_enable,
+					  bool remote_enable);
+
+/** @brief Set remote (peer) transmit power.
+ *
+ *  @param conn			Connection object.
+ *  @param phy			PHY bit number i.e. [1M, 2M, s8, s2] == [1, 2, 3, 4]
+ *  @param delta        Delta to be used by remote to apply to its transmit power.
+ *                      Delta is what we as local ask of the peer, it is not
+ *                      synonymous with the changes peer will (or won't) apply.
```
Comment (alwa-nordic) [inline include/zephyr/bluetooth/conn.h:606]: This sentence is complicated. Can we just call it a "request" instead?

File: include/zephyr/bluetooth/conn.h [hunk 2]
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
...
+ *                      Furthermore, it can be 0 to inquire tx power setting of remote.
```
Comment (alwa-nordic) [inline include/zephyr/bluetooth/conn.h:607]: How will the result be reported? I think we should note that here.

File: include/zephyr/bluetooth/conn.h [hunk 3]
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
...
+ *
+ *  @return Zero on success or (negative) error code on failure.
+ *  @return -ENOBUFS HCI command buffer is not available.
+ */
+int bt_conn_set_remote_tx_power_level(struct bt_conn *conn,
+					 enum bt_conn_le_tx_power_phy phy, int8_t delta);
+
+/** @brief Set remote (peer) transmit power.
```
Comment (alwa-nordic) [inline include/zephyr/bluetooth/conn.h:615]: This has the same description as `bt_conn_set_remote_tx_power_level`. Is that correct?

File: include/zephyr/bluetooth/conn.h [hunk 4]
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
...
+ *
+ *  @param conn              Connection object.
+ *  @param auto_enable       Enable or Disable controller initiated autonomous
+ *                           LE Power Control Request procedure.
+ *                           Disabled by default.
+ *  @param apr_enable        Enable or Disable APR handling in controller during
```
Comment (alwa-nordic) [inline include/zephyr/bluetooth/conn.h:621]: What's APR?

File: include/zephyr/bluetooth/conn.h [hunk 5]
```diff
@@ -597,6 +597,71 @@ int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
...
+ *                           LE Power Control Request procedure.
+ *                           Disabled by default.
+ *  @param beta              Beta value used for exponential weighted averaging of
```
Comment (alwa-nordic) [inline include/zephyr/bluetooth/conn.h:624]: Should we link https://en.wikipedia.org/wiki/Exponential_smoothing?

File: include/zephyr/bluetooth/conn.h [hunk 6]
```diff
@@ -597,6 +597,71 @@ int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
...
+ *                           The valid range of beta is [0, 4095].
+ *                           The beta value used in computation is beta/4096.
+ *                           For example, for beta to be 0.25 in the above computation,
+ *                           set the beta parameter in the command to 1024.
+ *                           Default value is 2048.
+ *  @param lower_limit       The lower limit of RSSI golden range, it is explained in
+ *                           Core_v5.3, Vol 6, Part B, Section 5.1.17.1, in dBm units.
+ *                           Default value is -70 dBm.
+ *  @param upper_limit       The upper limit of RSSI golden range, it is explained in
+ *                           Core_v5.3, Vol 6, Part B, Section 5.1.17.1, in dBm units.
```
Comment (alwa-nordic) [inline include/zephyr/bluetooth/conn.h:636]: The core spec reference is repeated a lot. I suggest we just define "golden range" once in the description.

File: include/zephyr/bluetooth/conn.h [hunk 7]
```diff
@@ -597,6 +597,71 @@ int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
...
+ *
+ *  @param conn              Connection object.
+ *  @param auto_enable       Enable or Disable controller initiated autonomous
+ *                           LE Power Control Request procedure.
+ *                           Disabled by default.
+ *  @param apr_enable        Enable or Disable APR handling in controller during
+ *                           LE Power Control Request procedure.
+ *                           Disabled by default.
+ *  @param beta              Beta value used for exponential weighted averaging of
+ *                           RSSI in 12-bit fixed point fraction.
```
Comment (alwa-nordic) [inline include/zephyr/bluetooth/conn.h:625]: It's not so clear what "12-bit fixed point fraction" means and it's confusing. Can we instead move the text "The beta value used in computation is beta/4096." up here?

**Review by alwa-nordic — COMMENTED** [review]
I'm not finished with the review. I'll continue tomorrow. You don't have to reply yet if you prefer me to finish first.

File: include/zephyr/bluetooth/conn.h [hunk 8]
```diff
@@ -479,6 +479,38 @@ struct bt_conn_le_tx_power {
...
+
+	/** Transmit power level */
+	int8_t tx_power_level;
+
+	/** 0: Transmit power level is at minimum level.
+	 *  1: Transmit power level is at maximum level.
+	 */
+	int8_t tx_power_level_flag;
+
+	/** Change in transmit power level */
```
[the remark this answers is not in the document]
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h]: Let me know if my formatting is good or has to be changed.

File: include/zephyr/bluetooth/conn.h [hunk 9]
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
...
+ */
+int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
+					  bool local_enable,
+					  bool remote_enable);
+
+/** @brief Set remote (peer) transmit power.
+ *
+ *  @param conn			Connection object.
+ *  @param phy			PHY bit number i.e. [1M, 2M, s8, s2] == [1, 2, 3, 4]
+ *  @param delta        Delta to be used by remote to apply to its transmit power.
```
[the remark this answers is not in the document]
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h:621]: Transmit power is in dBm, so an absolute value, and we request a delta of that, so the unit of delta is dBm... but might be wrong the sun has gone down hours ago (exaggeration) so my brain might be sleeping :D

File: include/zephyr/bluetooth/conn.h (hunk 1 above)
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h]: What do you think of the new sentences?

File: include/zephyr/bluetooth/conn.h (hunk 3 above)
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h]: whoops.

File: include/zephyr/bluetooth/conn.h (hunk 4 above)
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h]: Added into sentence, let me know if it is clear enough written.

File: include/zephyr/bluetooth/conn.h (hunk 5 above)
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h]: Is it normal to add wiki pages? And how to write it in that case? "see mathematics: wikipedia"? :D

File: include/zephyr/bluetooth/conn.h (hunk 6 above)
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h]: Or I change it like this. (I _really_ hope the comments pop up same spot on the new commit :D )

File: include/zephyr/bluetooth/conn.h (hunk 7 above)
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h:644]: Did so but it still seems a bit convoluted to me. Any ideas? I think it is most important to understand the gist of this... and anyways when I looked to understand this all books referencing this weren't available for free online, so anyways a bit difficult to find deeper understanding (and the algorithm was created by someone who left Nordic a while back)

File: include/zephyr/bluetooth/conn.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (alwa-nordic) [reply include/zephyr/bluetooth/conn.h:621]: Oh. Ok. I assumed it was the same kind of 'delta' as in Core 5.4 4.E.7.7.65.33. That has the unit dB.

image

File: include/zephyr/bluetooth/conn.h (hunk 7 above)
Comment (alwa-nordic) [reply include/zephyr/bluetooth/conn.h:644]: We can't document it as "Lost knowledge. Author quit." :P It has to mean something to our users.

I recognize `avg[n] = beta * avg[n - 1] + (1 - beta) * rssi[n]`. It's the mathematical form of the [Exponential Moving Average]. I assume `avg[n]` is then compared against the targets and limits.

It's not clear to me what `rssi[n]`. Is it the RX RSSI value observed at LL PDU number 'n'? Or the connection event number 'n'? Or something else?

'APR' is not a well known acronym. 'Acceptable power reduction'? Let's spell it out like 'Acceptable power reduction (APR)'. And what effect does that parameter have here?

`wait_period_ms` is also not clear to me. 

> Can we instead move the text "The beta value used in computation is beta/4096." up here?

> Did so but it still seems a bit convoluted to me. What is the trigger of the wait timeout?

I meant to suggest that we somehow replace the sentence "Beta value used for exponential weighted averaging of RSSI in 12-bit fixed point fraction." or remove it entirely. I think we should try to phrase the documentation in terms of what it means for a user of this API. I.e. What is the effect of various values for this parameter.

[Exponential Moving Average]: https://en.wikipedia.org/wiki/Exponential_smoothing

Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h:644]: Hi, I tried to make it more readable. I left the math, but renamed changed the beta/4096 to "gamma" in the description, so it is easier to understand: https://github.com/nrfconnect/sdk-zephyr/pull/1376/files#diff-4538e90f4e61dd34802aa1c0a5382dc89d4df6032ec6a3dbbf4eef746ebdd488R642
As per Teams communication I opened a PR in sdk-zephyr directly. If you want me to update this one as well, let me know.

**Pull request closed without merging**